# Imports

Imports -

In [2]:
%reload_ext autoreload
%autoreload 2

from book_functions import *
from signin_config import *

from tqdm import tqdm
import pandas as pd
import re
import requests
import signin_config
import pickle
import time
from bs4 import BeautifulSoup
from selenium import webdriver
from selenium.webdriver.common.by import By
from selenium.webdriver.common.keys import Keys
from selenium.common.exceptions import TimeoutException
from selenium.webdriver.support.ui import WebDriverWait
from selenium.webdriver.support import expected_conditions as EC
from imblearn.over_sampling import SMOTE
import time
from selenium.webdriver.chrome.options import Options

# Scraping goodreads

## Get intial 'top 200' urls list
Retrieve:


*continuous*: top rating, # of reviews, IBSN, GR_id#
      
*categorical*: author, image, top genre, description

--> add code for gathering list of top titles from past 50 years

In [5]:
url_list = get_urls()

## Use urls to get : title, authors, # of reviews and id

In [7]:
# test goodreads_list_scrape()

driver = webdriver.Chrome()
driver.get(url_list[0])
goodreads_list_scrape(driver);

In [21]:
# gather all initial book info

books_list = [] # list of dictionaries of each book

for url in tqdm(url_list):
    driver = webdriver.Chrome()
    driver.get(url)
    books_list += goodreads_list_scrape()
    driver.quit()

  0%|          | 0/48 [00:00<?, ?it/s]

50
100
150


  2%|▏         | 1/48 [00:23<18:44, 23.93s/it]

200
50
100
150


  4%|▍         | 2/48 [00:48<18:36, 24.27s/it]

200
50
100
150


  6%|▋         | 3/48 [01:12<18:03, 24.09s/it]

200
50
100
150


  8%|▊         | 4/48 [01:37<17:44, 24.20s/it]

200
50
100
150


 10%|█         | 5/48 [02:05<18:14, 25.45s/it]

200
50
100
150


 12%|█▎        | 6/48 [02:34<18:31, 26.47s/it]

200
50
100
150


 15%|█▍        | 7/48 [03:03<18:38, 27.27s/it]

200
50
100
150


 17%|█▋        | 8/48 [03:27<17:34, 26.35s/it]

200
50
100
150


 19%|█▉        | 9/48 [03:55<17:22, 26.74s/it]

50
100
150


 21%|██        | 10/48 [04:18<16:20, 25.81s/it]

200
50
100
150


 23%|██▎       | 11/48 [04:40<15:12, 24.67s/it]

200
50
100
150


 25%|██▌       | 12/48 [05:07<15:03, 25.11s/it]

200
50
100
150


 27%|██▋       | 13/48 [05:35<15:10, 26.02s/it]

200
50
100
150


 29%|██▉       | 14/48 [05:59<14:25, 25.45s/it]

200
50
100
150


 31%|███▏      | 15/48 [06:27<14:27, 26.28s/it]

200
50
100
150


 33%|███▎      | 16/48 [07:00<15:00, 28.13s/it]

200
50
100
150


 35%|███▌      | 17/48 [07:26<14:13, 27.53s/it]

200
50
100
150


 38%|███▊      | 18/48 [07:50<13:20, 26.69s/it]

200
50
100
150


 40%|███▉      | 19/48 [08:15<12:36, 26.09s/it]

200
50
100
150


 42%|████▏     | 20/48 [08:41<12:06, 25.95s/it]

200
50
100
150


 44%|████▍     | 21/48 [09:06<11:34, 25.73s/it]

200
50
100
150


 46%|████▌     | 22/48 [09:32<11:08, 25.71s/it]

200
50
100
150


 48%|████▊     | 23/48 [09:58<10:49, 25.98s/it]

200
50
100
150
200


 50%|█████     | 24/48 [10:24<10:20, 25.84s/it]

50
100
150


 52%|█████▏    | 25/48 [10:47<09:35, 25.04s/it]

200
50
100
150


 54%|█████▍    | 26/48 [11:13<09:18, 25.38s/it]

200
50
100
150


 56%|█████▋    | 27/48 [11:38<08:51, 25.32s/it]

200
50
100
150


 58%|█████▊    | 28/48 [12:03<08:20, 25.00s/it]

200
50
100
150


 60%|██████    | 29/48 [12:28<07:57, 25.15s/it]

200
50
100
150


 62%|██████▎   | 30/48 [12:57<07:51, 26.21s/it]

200
50
100
150


 65%|██████▍   | 31/48 [13:22<07:19, 25.85s/it]

200
50
100
150


 67%|██████▋   | 32/48 [13:49<06:59, 26.24s/it]

200
50
100
150


 69%|██████▉   | 33/48 [14:14<06:28, 25.88s/it]

200
50
100
150


 71%|███████   | 34/48 [14:41<06:08, 26.33s/it]

200
50
100
150


 73%|███████▎  | 35/48 [15:08<05:42, 26.34s/it]

200
50
100
150


 75%|███████▌  | 36/48 [15:37<05:28, 27.36s/it]

200
50
100
150


 77%|███████▋  | 37/48 [16:04<04:58, 27.14s/it]

200
50
100
150


 79%|███████▉  | 38/48 [16:32<04:32, 27.28s/it]

200
50
100
150


 81%|████████▏ | 39/48 [16:56<03:58, 26.49s/it]

200
50
100
150


 83%|████████▎ | 40/48 [17:20<03:26, 25.77s/it]

200
50
100
150


 85%|████████▌ | 41/48 [17:49<03:05, 26.49s/it]

200
50
100
150


 88%|████████▊ | 42/48 [18:16<02:41, 26.84s/it]

200
50
100
150


 90%|████████▉ | 43/48 [18:43<02:14, 26.89s/it]

200
50
100
150


 92%|█████████▏| 44/48 [19:06<01:42, 25.73s/it]

200
50
100
150


 94%|█████████▍| 45/48 [19:31<01:16, 25.33s/it]

200
50
100
150


 96%|█████████▌| 46/48 [19:57<00:51, 25.72s/it]

200
50
100
150


 98%|█████████▊| 47/48 [20:24<00:25, 25.98s/it]

200
50
100
150


100%|██████████| 48/48 [20:54<00:00, 26.13s/it]

200


Saving and loading via pickle:

In [10]:
# with open('bookslist_1909_on.pkl', 'wb') as f: 
#     pickle.dump(books_list, f)

# open stored bookslist if needed

with open('bookslist_1909_on.pkl', 'rb') as f:
    loaded_bookslist = pickle.load(f)

In [39]:
loaded_bookslist[:3]

NameError: name 'loaded_bookslist' is not defined

## Secondary scrape: Accessing img, description, format, pages and genre

In [30]:
# going to get info for all books, starting from 1909

# full_book_list = [] # <-- run if re-starting

driver = webdriver.Chrome()

initial_gr_signin(driver)

# for dic in books_list: if running whole notebook
for dic in loaded_bookslist:
    dic['img'],dic['description'], dic['format'], dic['page'], dic['genre'] = secondary_scrape(dic['id'], driver)        
    full_book_list.append(dic)

full_book_list[-1]

{'titles': 'To Catch an Heiress (Agents of the Crown, #1)',
 'authors': 'Paul Hoffman',
 'ratings': '3.90',
 'num_ratings': '12,269 ratings',
 'id': '110395',
 'img': '',
 'description': ' ',
 'format': '',
 'pages': '',
 'genre': set()}

In [134]:
len(full_book_list)

17727

## Get data into Pandas dataframe and save

Export dataframe as csv 

In [32]:
df1 = pd.DataFrame(full_book_list)
# df1 = df1.drop(['genre1', 'genre2'], axis=1)
# df1.to_csv('goodread_img1.csv')

In [40]:
df1.head()

titles               authors ratings  \
0  Anne of Avonlea (Anne of Green Gables, #2)       L.M. Montgomery    4.22   
1                                 Ethan Frome         Edith Wharton    3.39   
2                          The Arabian Nights             Anonymous    4.05   
3                                 Martin Eden           Jack London    4.43   
4   A Girl of the Limberlost (Limberlost, #2)  Gene Stratton-Porter    4.15   

  num_ratings      id                                                img  \
0      126781   77390  https://i.gr-assets.com/images/S/compressed.ph...   
1      100785    5246  https://i.gr-assets.com/images/S/compressed.ph...   
2       70660   93101  https://i.gr-assets.com/images/S/compressed.ph...   
3       21445  929782  https://i.gr-assets.com/images/S/compressed.ph...   
4       17932   17567  https://i.gr-assets.com/images/S/compressed.ph...   

                                         description       pages  \
0  At sixteen, Anne is grown up...almost. Her gra...   304 pages   
1  'He seemed a part of the mute melancholy lands...   189 pages   
2  The tales of told by Shahrazad over a thousand...  1049 pages   
3  Alternate cover edition of ISBN 9780140187724 ...   480 pages   
4  Set amid Indiana's vast Limberlost Swamp, this...   222 pages   

                                               genre format  
0  {'Childrens', 'Classics', 'Historical', 'Histo...    NaN  
1  {'Classics', 'Academic', 'Fiction', 'School', ...    NaN  
2  {'Classics', 'Short Stories', 'Fantasy', 'Fict...    NaN  
3  {'Classics', 'American', 'Literature', 'Fiction'}    NaN  
4  {'Childrens', 'Classics', 'Historical', 'Histo...    NaN

Import dataframe from csv

In [4]:
df1 = pd.read_csv('goodread_img1.csv')
df1.drop('Unnamed: 0', axis=1,inplace=True)
df1.head();

In [25]:
df = pd.read_csv('goodread_des1.csv')
df.drop('Unnamed: 0', axis=1,inplace=True)
df['isbn'] = np.nan

## Dealing with missing images

In [35]:
# see how many are empty
print('null:',df1.img.apply(lambda x: x if x != '' else np.nan).isnull().sum())
print ('not null:',df1.img.apply(lambda x: x if x != '' else np.nan).notnull().sum())

null: 211
not null: 17516


In [8]:
# create list with index & goodreads id for items missing an img source

no_img_id = []
for k, v in df1.iterrows():
    if df1.iloc[k].img == '' or type(df1.iloc[k].img) == float:
        no_img_id.append((k, df1.iloc[k].id))

len(no_img_id) # should be the same as above

10928

Testing get_imgs on individual site

In [10]:

driver = webdriver.Chrome()
initial_gr_signin(driver)
driver.get('https://www.goodreads.com/book/show/24950845')
img = ''
try:
    img = driver.find_element_by_id('coverImage')
    img = img.get_attribute('src')
    print ('cov')

except:
    try:
        img = driver.find_element_by_class('mainBookCover')
        img = img.get_attribute('src')
        print('main')

    except:
        try:
            img = driver.find_element_by_xpath('/html/body/div[2]/div[3]/div[1]/div[2]/div[2]/div[1]/div[1]/div[1]/div[1]/a/img')
            img = img.get_attribute('src')
            print('xpath')
        except:
            print('nope')
            img = np.nan

# igg = get_imgs(driver)
print(img)


cov
https://i.gr-assets.com/images/S/compressed.photo.goodreads.com/books/1431363154l/24950845.jpg


In [36]:
# check where in no_img_id the image search stopped

ct = 0
for i, ig in no_img_id:
    if i == 19161909 or ig == 19161909:
        print (ct)
    ct += 1

3116


In [36]:
len(df1), len(no_img_id), len(still_no_img)+found, found, len(still_no_img)

(10928, 10928, 10717, 211)

In [27]:
# code to get images using function

driver = webdriver.Chrome()
initial_gr_signin(driver)
# found = 0 
# still_no_img = []
skipped = 0

for idx, gr_id in no_img_id:
    if df1.iloc[idx].img != '' and type(df1.iloc[idx].img) != float:
        if skipped % 50 == 0:
            print (f'skipped : {skipped}')
        skipped +=1
        continue
    
    driver.get(f"https://www.goodreads.com/book/show/{gr_id}")
    
    img_src = get_imgs(driver)
    df1.loc[idx,'img'] = img_src
    
    if type(img_src) == float:
        print(f'image not found for {gr_id},')
        still_no_img.append((idx, gr_id))
    else:
        found += 1
    if found % 50 == 0 and found != 0 :
        print (f'found : {found}')
        
        # save updated dataframe, image list and found
        df1.to_csv('goodread_img1.csv')


image not found for 1069228,
found : 0
image not found for 5041959,
found : 0
image not found for 1350016,
found : 0
image not found for 2188544,
found : 0
image not found for 1740072,
found : 0
image not found for 1895199,
found : 0
image not found for 5962452,
found : 0
image not found for 5494385,
found : 50
image not found for 8701725,
image not found for 2634450,
image not found for 2571523,
image not found for 2579924,
image not found for 3278041,
image not found for 2624442,
image not found for 6651170,
image not found for 959528,
image not found for 1585444,
image not found for 6674903,
image not found for 2737795,
image not found for 6525886,
image not found for 2947536,
image not found for 2270923,
image not found for 2270839,
image not found for 3770082,
found : 100
found : 150
image not found for 6246317,
image not found for 107358,
image not found for 121328,
image not found for 2010401,
image not found for 3956198,
found : 200
image not found for 1839645,
image not found 

found : 7850
found : 7900
found : 7950
found : 8000
found : 8050
found : 8100
found : 8150
found : 8200
found : 8250
found : 8300
found : 8350
found : 8400
found : 8450
found : 8500
found : 8550
found : 8600
found : 8650
found : 8700
found : 8750
found : 8800
found : 8850
found : 8900
found : 8950
found : 9000
found : 9050
found : 9100
found : 9150
found : 9200
found : 9250
found : 9300
found : 9350
found : 9400
found : 9450
found : 9500
found : 9550
found : 9600
found : 9650
found : 9700
found : 9750
found : 9800
found : 9850
found : 9900
found : 9950
found : 10000
found : 10050
found : 10100
found : 10150
found : 10200
found : 10250
found : 10300
found : 10350
found : 10400
found : 10450
found : 10500
found : 10550
found : 10600
found : 10650
found : 10700


In [37]:
# final save with images (211 missing by the end)
df1.to_csv('goodread_img1.csv')

## Missing Description

Find where description is missing

In [5]:
no_description = []
for idx, row in df1[df1.description == ' '].iterrows():
    no_description.append((idx, row.id))


In [14]:
no_description[110]

(156, 5840903)

In [18]:
# testing get_description on individual site


driver = webdriver.Chrome()
initial_gr_signin(driver)
driver.get('https://www.goodreads.com/book/show/5840903')
descrip = ''

# more = driver.find_elements_by_xpath('//*[@id="description"]/a')
# try:
#     # handling errors to do with an extended description
#     more[0].click()
#     try:
#         describe = driver.find_element_by_id('description').text[:-7].strip()
#         descrip = ' '.join(describe.split('\n'))
#     except:
#         descrip = np.nan

# except:
#     try:
#         describe = driver.find_element_by_id('description').text[:-7].strip()
#         descrip = ' '.join(describe.split('\n'))
#     except:
#         descrip = np.nan

description = get_description(driver)
print (description)
# print(descrip, description, descrip == description)


When Jack receives a book on insects for his birthday, Maggie wishes one could get "into" a book. And so they do, experiencing adventures with insects of all sorts as they "travel" through the book. Along the way they encounter a butterfly with a grievance, extraordinary caterpillars, a hungry spider, a distinguished musician (a katydid), as well as fireflies, locusts, grasshoppers, wasps, bees, and more. The author, Edmund Selous, a lifelong observer of wildlife, produced books about animals for both children and adults. In his books for children, Selous conveys an amazing amount of information about animals through the conversations his youthful protagonists have with the animals they encounter. Each animal is brought to life so vividly through the story that the reader unconsciously absorbs the details of its existence: its physical characteristics, its habits and preferences, and its interactions with other animals.
 When Jack receives a book on insects for his birthday, Maggie wis

In [22]:
df1.to_csv('goodread_des1.csv')

In [27]:
count = 1
for i, ig in no_description:
    if count == 2447:
        print (i, ig)
    count += 1

2605 15819761


In [21]:
# code to get description using function

driver = webdriver.Chrome()
initial_gr_signin(driver)
found = 0 
still_no_des = []
skipped = 0

for idx, gr_id in no_description:
    if df1.iloc[idx].description != ' ' and type(df1.iloc[idx].description) != float:
        continue
    
    driver.get(f"https://www.goodreads.com/book/show/{gr_id}")
    
    description = get_description(driver)
    df1.loc[idx,'description'] = description
    
    if type(description) == float:
        print(f'description not found for {gr_id},')
        still_no_des.append((idx, gr_id))
    else:
        found += 1
    if found % 50 == 0 and found != 0 :
        print (f'found : {found}')
        
        # save updated dataframe, image list and found
        df1.to_csv('goodread_des1.csv')


description not found for 6480591,
description not found for 483583,
found : 50
found : 100
description not found for 2567639,
description not found for 6651170,
description not found for 3245210,
description not found for 1585444,
description not found for 2737795,
description not found for 6525886,
description not found for 2270839,
description not found for 3770082,
found : 150
description not found for 2727141,
found : 200
description not found for 107358,
description not found for 7705411,
description not found for 5263927,
description not found for 1346496,
description not found for 2154455,
found : 250
description not found for 1003199,
found : 250
description not found for 1333884,
description not found for 782909,
description not found for 2435094,
description not found for 2200868,
description not found for 6830705,
description not found for 1427509,
description not found for 3891281,
description not found for 2348468,
found : 300
found : 350
found : 400
found : 450
descripti

KeyboardInterrupt: 

makes sense: 6480591 , 483583 

What's up with : Henry James?

## Missing Genres

In [3]:
# run at onset of df creation, if scraping from the top

df['au_ge'] = ''
df['desc_clean'] = ''
df['key_words'] = ''
df['bag_of_words'] = ''

SyntaxError: invalid syntax (<ipython-input-3-7eee880c0f25>, line 6)

In [36]:
# code to get form,page,isbn,genre and desc using functions

driver = webdriver.Chrome()
initial_gr_signin(driver)
found = 0 
no_form = []
no_page = []
no_isbn = []
no_des = []
skipped = 0


for idx, row in df.iloc[start:].iterrows():
    driver.get(f"https://www.goodreads.com/book/show/{row.id}")
    form, page, isbn = get_form_page_isbn(driver)
    genre = get_genre(driver)
    
    df.loc[idx,'format'] = form
    df.loc[idx,'pages'] = page
    df.loc[idx,'isbn'] = isbn
    df.loc[idx,'genre'] = genre
    
    description = get_description(driver)
    df.loc[idx,'description'] = description
    if type(description) == float:
        no_des.append((idx, row.id))

    if type(form) == float:
        no_form.append((idx, row.id))
    elif type(page) == float:
        no_page.append((idx, row.id))
    elif type(isbn) == float:
        no_isbn.append((idx, row.id))
    elif type(genre) == float:
        no_genre.append((idx, row.id))

    found += 1
        
    if found % 50 == 0:
        df_lite = df.iloc[start:end]
        print (f'found all : {found}')
        
        clean_bag_save(df_lite, 'goodreads_updated')

        start += 50


df.to_csv('goodreads_updated.csv')
df.to_json(r'goodreads_updated.json',orient='index')


found all : 50
found all : 100
found all : 100
found all : 150
found all : 200
found all : 250
found all : 250
found all : 300
found all : 350
found all : 400
found all : 450
found all : 450
found all : 500
found all : 550
found all : 600
found all : 600
found all : 650
found all : 700
found all : 700
found all : 750
found all : 800
found all : 850
found all : 900
found all : 900
found all : 950
found all : 1000
found all : 1000
found all : 1050
found all : 1100
found all : 1150
found all : 1200
found all : 1200
found all : 1250
found all : 1250
found all : 1300
found all : 1300
found all : 1350
found all : 1400
found all : 1450
found all : 1500
found all : 1550
found all : 1600
found all : 1650
found all : 1700
found all : 1750
found all : 1800
found all : 1850
found all : 1900
found all : 1950
found all : 1950
found all : 1950
found all : 1950
found all : 2000
found all : 2000
found all : 2000
found all : 2000
found all : 2050
found all : 2100
found all : 2150
found all : 2200
found 

KeyboardInterrupt: 

In [37]:
df.to_csv('goodreads_updated.csv')

# Data Cleaning

## Clean data 

- look for nan values, empty descriptions, clean types (str -> int), (organize pictures)
    ratings, number of ratings, pages (str -> int)

Add bag of words, au_ge and keywords

Process and create bag of words

In [40]:
au_ge, keywords = process_and_almost_bag(df.authors, df.genre, df.desc_clean)
boow = make_bow(aug, kee)

'hi bye'